x/ccv/provider/keeper/relay.go

```diff
@@ -112,14 +100,33 @@ func (k Keeper) EndBlockCallback(ctx sdk.Context) {
 	// get the validator updates from the staking module
 	valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)
 	k.IterateConsumerChains(ctx, func(ctx sdk.Context, chainID string) (stop bool) {
+		// check whether there is an established CCV channel to this consumer chain
+		if channelID, found := k.GetChainToChannel(ctx, chainID); found {
+			// send all the pending ValidatorSetChangePackets to the consumer chain
+			pendingPackets := k.EmptyPendingVSC(ctx, chainID)
+			for _, data := range pendingPackets {
+				k.SendValidatorSetChangePacket(ctx, channelID, data)
+			}
+		}
+
 		// check whether there are changes in the validator set;
 		// note that this also entails unbonding operations
 		// w/o changes in the voting power of the validators in the validator set
 		unbondingOps, _ := k.GetUnbondingOpsFromIndex(ctx, chainID, valUpdateID)
 		if len(valUpdates) != 0 || len(unbondingOps) != 0 {
-			k.SendValidatorSetChangePacket(ctx, chainID, valUpdates, valUpdateID, k.EmptySlashAcks(ctx, chainID))
+			// construct validator set change packet data
+			packetData := ccv.NewValidatorSetChangePacketData(valUpdates, valUpdateID, k.EmptySlashAcks(ctx, chainID))
+
+			// check whether there is an established CCV channel to this consumer chain
+			if channelID, found := k.GetChainToChannel(ctx, chainID); found {
+				// send this validator set change packet data to the consumer chain
+				k.SendValidatorSetChangePacket(ctx, channelID, packetData)
+			} else {
+				// store the packet data to be sent once the CCV channel is established
+				k.AppendPendingVSC(ctx, chainID, packetData)
```

Comment:
Seems like we can greatly simplify this code if we resolve this issue: https://github.com/cosmos/ibc/issues/645

I can take that up.

Reply:
See my comment below https://github.com/cosmos/interchain-security/pull/163#issuecomment-1168389102